src/daft-plan/src/logical_ops/unpivot.rs

```diff
@@ -0,0 +1,94 @@
+use std::sync::Arc;
+
+use common_error::{DaftError, DaftResult};
+use daft_core::{
+    datatypes::Field,
+    schema::{Schema, SchemaRef},
+    utils::supertype::try_get_supertype,
+    DataType,
+};
+use daft_dsl::ExprRef;
+
+use itertools::Itertools;
+use snafu::ResultExt;
+
+use crate::{
+    logical_plan::{self, CreationSnafu},
+    LogicalPlan,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Unpivot {
+    pub input: Arc<LogicalPlan>,
+    pub ids: Vec<ExprRef>,
+    pub values: Vec<ExprRef>,
+    pub variable_name: String,
+    pub value_name: String,
+    pub output_schema: SchemaRef,
+}
+
+impl Unpivot {
+    pub(crate) fn try_new(
+        input: Arc<LogicalPlan>,
+        ids: Vec<ExprRef>,
+        values: Vec<ExprRef>,
+        variable_name: &str,
+        value_name: &str,
+    ) -> logical_plan::Result<Self> {
+        if values.is_empty() {
+            return Err(DaftError::ValueError(
+                "Unpivot requires at least one value column".to_string(),
+            ))
+            .context(CreationSnafu);
+        }
+
+        let input_schema = input.schema();
+        let values_fields = values
+            .iter()
+            .map(|e| e.to_field(&input_schema))
+            .collect::<DaftResult<Vec<_>>>()
+            .context(CreationSnafu)?;
+
+        let value_dtype = values_fields
+            .iter()
+            .map(|f| f.dtype.clone())
+            .try_reduce(|a, b| try_get_supertype(&a, &b))
```

Comment:
i feel like there's a lot of try_get_super_type calls, would it make sense to just do it once, then pipe the type down all the way into the actual unpivot logic?

Reply:
Yeah I agree with you. The thing is both the logical plan and the micropartition expect a schema, but micropartitions can also be created on their own, which means it won't make sense to pipe the result from the logical op creation. Moreover I also do schema resolution differently depending on if the micropartition is empty, but it is trivial to derive a schema from a table so I don't see a great reason to combine them. 

But that means there are three places in the code where we find the dtype of the value. I'm open to any ideas about how to improve that!

Reply:
Ah gotcha, I think it's fine then.